data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/chance_to_hit/chance_to_hit_high_fraction.cfg

```diff
@@ -0,0 +1,39 @@
+# wmllint: no translatables
+
+#####
+# API(s) being tested: [chance_to_hit]value=
+##
+# Actions:
+# Give the leaders a chance_to_hit ability with a value of 99.8.
```

Comment:
That seems to be a 1.8% failure chance, there's probability (0.998 ^ 2000) that all 2000 strikes hit.

Reply:
It truncates the value to 99%, so it's not `.998^1000 = 0.135064522` it's `.99^1000 = 0.000043171`.

Reply:
The comment should be much clearer about what the test is actually expecting, if it's expecting 99.8 to be equivalent to 99 then it should say that.

The failure rate is now 1/10000, and another 1/10000 for chance_to_hit_low_fraction;  if we're going to have them at all then they should have warnings in the comments that this method is only for the CTH tests,and shouldn't be copied elsewhere. It feels wrong to have non-deterministic unit tests, particularly in a game that includes small probabilities adding up; I'd still much rather have the code use simulate_combat.

Reply:
I can increase the strikes count to make failure less likely, but using simulate_combat defeats the purpose - these are supposed to be a test of what actually happens, not a test of what simulate_combat does.

Reply:
If we document that the value is truncated, then why would UMC authors use a value with a decimal in anyway? I think we should just test that a number with a decimal in goes to a reasonable value, but it can be a much wider range than checking that 99% CTH doesn't hit 1000/1000 times. Given how long the animation takes, I think there should never be a time when a player encounters 99% CTH - either use 100%, or use something that gives a reasonable risk of missing.

I see a reason to test that, if a WML uses a fraction, they get something reasonably close to what they wanted, for example that `value=80.1` doesn't turn into 0%, 100%, or the terrain defense that the unit is standing on. For testing purposes, I suggest checking `value=80.8` and 1000 strikes results in between 650 and 950 hits. That's checks the effective CTH is neither 40% (the default from the castle that Bob is standing on), nor 100%, while still being practically deterministic (somewhere around 1 in 10^26 chance of failure).

The stuff below is a smaller improvement written before the paragraphs above:
```
#####
# API(s) being tested: [chance_to_hit]value=
##
# Actions:
# Give the leaders a chance_to_hit ability with a value of 99.999.
# Have the side 1 leader attack the side 2 leader with both its weapons
##
# Expected end state:
# The value is truncated to an integer percentage, so value=99.999 means only 99%.
# The side 1 leader's first weapon strikes less than 1000 times
# The side 1 leader's second weapon strikes less than 1000 times
#####
#ifndef SCHEMA_VALIDATION
{COMMON_KEEP_A_B_UNIT_TEST "chance_to_hit_high_fraction_nondeterministic" (
    [event]
        name = start

        {SET_HP VALUE=10000}

        [modify_unit]
            [filter]
            [/filter]
            [effect]
                apply_to = new_ability
                [abilities]
                    {TEST_ABILITY attacks 1000 () SELF=yes}
                    {TEST_ABILITY chance_to_hit 99.999 () SELF=yes}
                [/abilities]
            [/effect]
        [/modify_unit]
    [/event]

    # There is a small chance of non-deterministic failure here. Using simulate_combat would defeat
    # the purpose of the test, which is to check that the truncated value is used in the fight.
    # TODO say what the actual probability is
    #
    # TODO Is there a value to this test? Checking that the value is used could be done with a value
    # around 80.8%, and a test that the result is within 65-95% - as we expect the value to be
    # truncated, WML authors shouldn't be adding decimal places anyway.
    {CHECK_STRIKES "1000,1000" ({SUCCEED}) COMPARE=less_than (CHANCE_TO_HIT=)}
)}
#endif
```

Reply:
I'm not really sure what the issue is - literally every other set of tests I've written for abilities that take a value has had both these tests to show that fractional values are truncated. If you don't think this needs to be tested for chance_to_hit because the effect of rounding up vs down doesn't realistically matter, then I'll just remove these two tests.

Reply:
The main issue is that this test isn't deterministic, and it's thus contributing to the set of CI runs that give misleading results. All the other ones are deterministic.

Reply:
I will removed these two tests then.